src/Aspire.Hosting/Dcp/Model/Container.cs

```diff
@@ -283,6 +287,146 @@ internal sealed class ContainerPortSpec
+        {
+            Type = type,
+            Name = item.Name,
+            Owner = item.Owner,
+            Group = item.Group,
+            Mode = (int)item.Mode,
+        };
+
+        if (item is ContainerFile file)
+        {
+            entry.Contents = file.Contents;
+        }
+        else if (item is ContainerDirectory directory)
+        {
+            entry.Entries = directory.Entries?.Select(e => e.ToContainerFileSystemEntry()).ToList();
+        }
+
+        return entry;
+    }
+}
+
+internal sealed class ContainerFileSystemEntry
+{
+    // The type of the file system entry (file or directory)
+    [JsonPropertyName("type")]
+    public string Type { get; set; } = ContainerFileSystemEntryType.File;
+
+    // The name of the file system entry
+    [JsonPropertyName("name")]
+    public string? Name { get; set; }
+
+    // The UID of the user that owns the file system (if null, the owner will be inherited from its parent directory or the request default)
+    [JsonPropertyName("owner")]
+    public int? Owner { get; set; }
+
+    // The GID of the group that owns the file system (if null, the group will be inherited from its parent directory or the request default)
+    [JsonPropertyName("group")]
+    public int? Group { get; set; }
+
+    // The file system mode (permissions) of the file system entry (if 0, the mode will be inherited from its parent directory or the request default)
+    [JsonPropertyName("mode")]
+    public int Mode { get; set; }
+
+    // If the file system entry is a file, this is the contents of that file. Setting Contents for a directory is an error.
+    [JsonPropertyName("contents")]
+    public string? Contents { get; set; }
+
+    // If the file system entry is a directory, this is the list of entries in that directory. Setting Entries for a file is an error.
+    [JsonPropertyName("entries")]
+    public List<ContainerFileSystemEntry>? Entries { get; set; }
+}
+
+internal sealed class ContainerCreateFileSystemComparer : IEqualityComparer<ContainerCreateFileSystem>
+{
+    public bool Equals(ContainerCreateFileSystem? x, ContainerCreateFileSystem? y)
+    {
+        if (x is null || y is null)
+        {
+            return false;
+        }
+
+        return x.Destination == y.Destination
+            && x.DefaultOwner == y.DefaultOwner
+            && x.DefaultGroup == y.DefaultGroup
+            && x.Mode == y.Mode
+            && (x.Entries?.SequenceEqual(y.Entries ?? Enumerable.Empty<ContainerFileSystemEntry>(), new ContainerFileSystemEntryComparer()) ?? false);
+    }
+
+    public int GetHashCode(ContainerCreateFileSystem obj)
+    {
+        return HashCode.Combine(obj.Destination, obj.DefaultOwner, obj.DefaultGroup, obj.Mode);
+    }
+}
+
+internal sealed class ContainerFileSystemEntryComparer : IEqualityComparer<ContainerFileSystemEntry>
+{
+    public bool Equals(ContainerFileSystemEntry? x, ContainerFileSystemEntry? y)
+    {
+        if (x is null || y is null)
```

Comment:
Is .NET convention to return false if both operands are null?

Reply:
yes